pkg/discovery/processor/cluster_processor.go

```diff
@@ -2,31 +2,85 @@ package processor
 
 import (
 	"fmt"
+	"github.com/turbonomic/kubeturbo/pkg/features"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
 	"time"
 
+	"github.com/KimMachineGun/automemlimit/memlimit"
 	"github.com/golang/glog"
+	v1 "k8s.io/api/core/v1"
+
 	"github.com/turbonomic/kubeturbo/pkg/cluster"
 	"github.com/turbonomic/kubeturbo/pkg/discovery/repository"
+	nodeUtil "github.com/turbonomic/kubeturbo/pkg/discovery/util"
 	"github.com/turbonomic/kubeturbo/pkg/kubeclient"
-	v1 "k8s.io/api/core/v1"
+)
+
+const (
+	// DefaultItemsPerGBMemory defines number of items to retrieve for each GB of memory
+	DefaultItemsPerGBMemory                     = 5000
+	DefaultMemoryLimitInGB                      = 8
+	DefaultAutoMemLimitPct                      = 0.9
+	DefaultGenericMetricSizePerThousandPodsInGB = 0.24
+	DefaultNumberOfPodsInThousands              = 1
+	DefaultCadvisorMetricSizePerPodInGB         = 0.004
+	DefaultMaxPodsPerNode                       = 250
+	DefaultExtraPerNodeUsageInGB                = 0.2
+	DefaultExtraClusterWideUsageInGB            = 0.2
 )
 
 var (
 	workers       = 10
 	totalWaitTime = 60 * time.Second
+	// DefaultItemsPerListQuery defines default number of items that should be requested in each
+	// workload controller list API to ensure no OOMs occur.
+	//
+	// This value is calculated using the following expression:
+	//
+	// (kubeturbo_mem_limit_gb * default_automemlimit_pct
+	//    - generic_metric_size_gb_per_thousand_pods * number_of_pods_in_thousands_in_cluster
+	//    - (cadvisor_metric_size_gb_per_pod * max_pods_per_node + extra_per_node_usage_gb)
+	//    - extra_cluster_wide_usage_gb) * items_per_gb
+	//
+	// In the absence of any MEMLIMIT set, we assume the following default values
+	// to calculate the default items per query:
+	//   * mem_limit_gb: 8
+	//   * default_automemlimit_pct: 0.9 (percentage of memory limit that can be controlled by Go Runtime)
+	//   * generic_metric_size_gb_per_thousand_pods: 0.24
+	//   * number_of_pods_in_thousands_in_cluster: 1 (1000 pods)
+	//   * cadvisor_metric_size_gb_per_pod: 0.004
+	//   * max_pods_per_node: 250
+	//   * extra_per_node_usage_gb: 0.2
+	//   * extra_cluster_wide_usage_gb: 0.2
+	//   * items_per_gb: 5000
+	//
+	// which results in the value of:
+	//   (8 * 0.9 - 0.24 * 1 - (0.004 * 250 + 0.2) - 0.2) *  5000 = 27800
+	//
+	// This default is unlikely to be used on Linux based system as Cgroup limit will always
+	// be available. If the container limit is not set, Cgroup limit will be set to nodes limit.
```

Comment:
I wonder if it'd be too aggressive to use the node memory limit to compute the items per query.  Perhaps we should use the half of the node limit instead, assuming there are others on the node too which could assume a significant amount of memory.  We can consider doing that in a subsequent PR, and leave the current one as is.

Reply:
From @ading1977 

Regarding halve the node limit, the challenge is that it is not straightforward to tell if the limit is actually a node limit, or a user specified container limit …

From @chlam4 

I think regardless it’s user specified or inherited from the node, it presents a higher risk if the container limit is too close to the node’s (unless its request is also as high).  This may sound like crazy, but a formula like the following to adjust the container limit could work well in practice:
`effective_container_limit = container_limit * ( 1 - 2 ^ ((container_limit - container_request) / node_limit) )`
As this is definitely crazy and probably not easy to explain, we can just leave as is and see what happens in practice.  Perhaps the 2x buffer is enough to minimize any OOMs.